lib/ecto/query.ex

```diff
@@ -2262,6 +2262,49 @@ defmodule Ecto.Query do
     has_named_binding?(Ecto.Queryable.to_query(queryable), key)
   end
 
+  @doc """
+  Applies the callback function to the query if it does not have a named binding,
+  otherwise returns the query without changes.
+
+  The function accepts an `Ecto.Query` struct and must return the query with an added 
+  named binding.
+
+  ## Examples
+
+      with_named_binding(query, :comments, fn query ->
+        join(query, :left, c in assoc(p, :comments), as: :comments)
+      end)
+
+  For more information on named bindings see "Named bindings" in this module doc.
+  """
+  def with_named_binding(%Ecto.Query{} = query, key, fun) when is_function(fun, 1) do
+    if has_named_binding?(query, key) do
+      query
+    else
+      query
+      |> fun.()
+      |> raise_on_invalid_callback_return(key)
+    end
+  end
+
+  def with_named_binding(queryable, key, fun) do
```

Comment:
Please keep me honest, but if it's not a query already then there shouldn't be any uncertainty whether the binding exists and this function wouldn't be needed?

Reply:
There is one edge case why I left it as is: you can implement `Ecto.Queryable` protocol for any struct, protocol implementation can do pretty much anything it wants, including creating aliases. So it's a not safe to assume that `to_query/1` call will always return an `Ecto.Query` without aliases. (However, I agree it's highly unlikely as I have never seen such use cases.)

Plus the code in `has_named_binding?/3` can change and I feel generally safer to rely on it rather than duplicating the logic.